src/utils/src/ExponentialBackoffUtils.c

```diff
@@ -0,0 +1,223 @@
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pExponentialBackoffState != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffState->currentRetryCount = 0;
+    pExponentialBackoffState->lastRetryWaitTime = 0;
+    pExponentialBackoffState->status = BACKOFF_NOT_STARTED;
+    DLOGD("Resetting Exponential Backoff State");
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS createExponentialBackoffStateWithDefaultConfig(PExponentialBackoffState* ppExponentialBackoffState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PExponentialBackoffState pExponentialBackoffState = NULL;
+    PExponentialBackoffConfig pExponentialBackoffConfig = NULL;
+
+    CHK(ppExponentialBackoffState != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffState = (PExponentialBackoffState) MEMCALLOC(0x00, SIZEOF(ExponentialBackoffState));
+    CHK(pExponentialBackoffState != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    CHK_STATUS(getDefaultExponentialBackOffConfig(&pExponentialBackoffConfig));
+
+    pExponentialBackoffState->pExponentialBackoffConfig = pExponentialBackoffConfig;
+    CHK_STATUS(resetRetryState(pExponentialBackoffState));
+
+    CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppExponentialBackoffState) = pExponentialBackoffState;
+    } else {
+        if (pExponentialBackoffState != NULL) {
+            SAFE_MEMFREE(pExponentialBackoffState->pExponentialBackoffConfig);
+            SAFE_MEMFREE(pExponentialBackoffState);
+        }
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS createExponentialBackoffState(PExponentialBackoffState* ppBackoffState, PExponentialBackoffConfig pBackoffConfig) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PExponentialBackoffState pExponentialBackoffState = NULL;
+    PExponentialBackoffConfig pExponentialBackoffConfig = NULL;
+
+    CHK(ppBackoffState != NULL, STATUS_NULL_ARG);
+    CHK(pBackoffConfig != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffState = (PExponentialBackoffState) MEMCALLOC(0x00, SIZEOF(ExponentialBackoffState));
+    CHK(pExponentialBackoffState != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    pExponentialBackoffConfig = (PExponentialBackoffConfig) MEMCALLOC(0x00, SIZEOF(ExponentialBackoffConfig));
+    CHK(pExponentialBackoffConfig != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    *pExponentialBackoffConfig = *pBackoffConfig;
+
+    pExponentialBackoffState->pExponentialBackoffConfig = pExponentialBackoffConfig;
+    CHK_STATUS(resetRetryState(pExponentialBackoffState));
+
+    CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppBackoffState) = pExponentialBackoffState;
+    } else {
+        SAFE_MEMFREE(pExponentialBackoffState);
+        SAFE_MEMFREE(pExponentialBackoffConfig);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+UINT64 getRandomJitter(UINT32 jitterFactor) {
+    return (RAND() % jitterFactor) * HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+}
+
+UINT64 power(UINT32 base, UINT32 exponent) {
+    UINT64 result = 1;
```

Comment:
Ok, as this is a generic API, you might want to sanitize/validate the input either here or in the function that calls this.

Reply:
done